### ISSUE TITLE
add a better way to attach a png to a map plot

### ISSUE THREAD

**Issue by danlipsa** [UV-CDAT/vcs#162]
Currently, put_png is used to put a background image on the canvas. A template is used to scale the png image so that it fits the map plot, so the user has to fidget with the scaling option to get it to fit.

Instead, the scaling can be computed vcs so that the png fits the map plot. It can also be extended to be used in map plots that use geographic projections.

Using the new API a user would only have to specify a png as an option to the plot. The image would be scaled so that it fits with the plot.

@doutriaux1 @aashish24 @durack1 What do you think about this enhancement?

**Comment by durack1** [comment]
@danlipsa this would be unreal! I wonder if the png hard code is optional? Could this also work with tif, jpeg etc inputs and overlays?

**Comment by danlipsa** [comment]
@durack1 The png file format would not be  required. It should work with any image known to VTK.

**Comment by doutriaux1** [comment]
@danlipsa I think that would be perfect! How long do you think it would take?

**Comment by danlipsa** [comment]
@doutriaux1 A rough estimate: a few days to a week to add to one plot and maybe a few more days to extended to other plots.

**Comment by danlipsa** [comment]
@durack1 I am nor sure what do you have in mind.

**Comment by doutriaux1** [comment]
ok, it's neat but I would vote to fix the test suite first on circleci.

**Comment by doutriaux1** [comment]
we can always put it in a nightly after the fact for @durack1

**Comment by danlipsa** [comment]
@doutriaux1 Yes, definitely, we'll do this after the release - And we cannot release unless we have a working test suit.

**Comment by durack1** [comment]
@doutriaux1 @danlipsa `nightly` is fine with me... I could even test out a branch if it was available.. With something like this user feedback could be helpful to round out the edges

**Comment by durack1** [comment]
@danlipsa, regarding **other** stuff to expose.. I'm naive as to what VTK can do, so maybe I need to browse the VTK example gallery and see what else is possible

**Comment by danlipsa** [comment]
@durack1 @aashish24 @doutriaux1 Unfortunately, we are not as good as we should be in showcasing all visualizations we support. So, if you see something you like bring it up and we'll look into it. It is quite possible that we already support it.
Our gallery is focused on 3D and is not comprehensive.
http://www.vtk.org/gallery/

You can also take a look at 
http://www.vtk.org/features/
2D Plots and Charts and Infovis would be of interest, but again those are not comprehensive.

**Comment by durack1** [comment]
@danlipsa thanks for the links.. I just took a peek and there was nothing there that jumped out at me.. Your suggestion of raising a query with an example (or sketch or similar) will be something I'll consider in the future..

**Comment by durack1** [comment]
@danlipsa with some considerable trial and error and a lot of help from @doutriaux1 I have managed to get the image that I was imagining:
![170417_pcmdi-amip-1-1-2_siconcandtos-201612](https://cloud.githubusercontent.com/assets/3229632/25148790/016d8b18-2431-11e7-9dca-eb6992f949ab.png)
This required some considerable kludging to get the desired output, but the final product does look clean. I think this provides a good template to aim toward when addressing this issue

**Comment by danlipsa** [comment]
@durack1 @aashish24 @doutriaux1 Very nice! I think it would be great to get something like this without too much effort.

**Comment by durack1** [comment]
@danlipsa @doutriaux1 @williams13 what would be **really** nice, is the ability to georeference the `png` overlay/input file so that you could rearrange it however you wanted, so for the lower panel above becomes e.g.:
![170420_pcmdi-amip-1-1-2_siconcandtos-201612_completepacific](https://cloud.githubusercontent.com/assets/3229632/25402463/9fcf3460-29ad-11e7-90d7-b9ed15f26c17.png)
This would also further augment the functionality so you could potentially use maps that are not global, so the ability to georeference the overlay, and maybe even many of them, if you have a mosaic of high res files for each continent for e.g.
As I noted before I am more than happy to iterate with your guys when this issue is starting to be looked at

**Comment by danlipsa** [comment]
@durack1 Yes, definitely this is how we'll do it. I think there are images that are already geo-referenced. @aashish24 would know more about this.

**Comment by durack1** [comment]
@danlipsa let me know when you get to look at this, we have a pub in the revision queue and I would really like to highlight this CDAT functionality if it's available at publication time..

**Comment by danlipsa** [comment]
@durack1 That would be very nice. I have vcs.js and upgrading vcs to OpenGl2 in my queue. Keep promoting this to @williams13 @doutriaux1 and @aashish24 and will get done.

**Comment by durack1** [comment]
@danlipsa @doutriaux1 @williams13 @aashish24 I would certainly up-vote this feature, as it's not obvious to me that any other package is capable of generating these

**Comment by durack1** [comment]
For a matplotlib example, with multiple options (blue marble, topographic relief etc) see https://matplotlib.org/basemap/users/geography.html

**Comment by doutriaux1** [comment]
@durack1 yes but these are canned images by the developers, correct? We want full featured stuff, with any image.

**Comment by lee1043** [comment]
@doutriaux1 @durack1 Yes, it is canned image, but key thing of it is that the map image can be fitted to any type of map projection or region, automatically.

**Comment by durack1** [comment]
The actual function is `warpimage()` ***which is only global*** - see: use an abitrary image as a map background. The image must be global, covering the world in lat/lon coordinates from the international dateline eastward and the South Pole northward.

So there is considerable scope to do things alot better here!

**Comment by jypeter** [comment]
There are also examples of what you can do with cartopy. Check the _Web services_ and _Miscellanea_ sections of http://scitools.org.uk/cartopy/docs/latest/gallery.html

But these examples are getting their data from external map services, not from a local png file. But maybe you could use a local file by using their [shapereader](http://scitools.org.uk/cartopy/docs/latest/tutorials/using_the_shapereader.html). There is also this [raster image ](http://scitools.org.uk/cartopy/docs/latest/developer_interfaces.html#raster-images) stuff. I confess I'm not an expert at all
